File: estateman-frontend/estateman-backend/app/api/v1/endpoints/properties.py
from fastapi import APIRouter, Depends, HTTPException, Query, UploadFile, File
from sqlalchemy.orm import Session
from typing import List, Optional
from datetime import datetime
import os
import uuid
from ....core.database import get_db
from ....services.property import PropertyService
from ....api.deps import get_current_user
from ....models.user import User
from pydantic import BaseModel

router = APIRouter()

class PropertyCreate(BaseModel):
    title: str
    description: Optional[str] = None
    property_type: str
    price: float
    bedrooms: Optional[int] = None
    bathrooms: Optional[int] = None
    square_feet: Optional[float] = None
    lot_size: Optional[float] = None
    year_built: Optional[int] = None
    address: str
    city: str
    state: str
    zip_code: str
    latitude: Optional[float] = None
    longitude: Optional[float] = None

class PropertyUpdate(BaseModel):
    title: Optional[str] = None
    description: Optional[str] = None
    property_type: Optional[str] = None
    status: Optional[str] = None
    price: Optional[float] = None
    bedrooms: Optional[int] = None
    bathrooms: Optional[int] = None
    square_feet: Optional[float] = None
    lot_size: Optional[float] = None
    year_built: Optional[int] = None
    address: Optional[str] = None
    city: Optional[str] = None
    state: Optional[str] = None
    zip_code: Optional[str] = None
    latitude: Optional[float] = None
    longitude: Optional[float] = None

class PropertyImageCreate(BaseModel):
    image_url: str
    caption: Optional[str] = None
    is_primary: bool = False
    order_index: int = 0

class PropertyDocumentCreate(BaseModel):
    document_url: str
    document_name: str
    document_type: Optional[str] = None
    file_size: Optional[int] = None

class PropertyShowingCreate(BaseModel):
    client_name: str
    client_email: Optional[str] = None
    client_phone: Optional[str] = None
    showing_date: datetime
    duration_minutes: int = 30
    notes: Optional[str] = None

class PropertyShowingUpdate(BaseModel):
    status: Optional[str] = None
    feedback: Optional[str] = None
    notes: Optional[str] = None

@router.get("/")
async def get_properties(
    skip: int = 0,
    limit: int = 100,
    status: Optional[str] = None,
    property_type: Optional[str] = None,
    min_price: Optional[float] = None,
    max_price: Optional[float] = None,
    city: Optional[str] = None,
    bedrooms: Optional[int] = None,
    bathrooms: Optional[int] = None,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    property_service = PropertyService(db)
    filters = {
        "status": status,
        "property_type": property_type,
        "min_price": min_price,
        "max_price": max_price,
        "city": city,
        "bedrooms": bedrooms,
        "bathrooms": bathrooms
    }
    filters = {k: v for k, v in filters.items() if v is not None}
    
    properties = property_service.get_properties(skip=skip, limit=limit, filters=filters)
    return {"properties": properties or []}

@router.get("/{property_id}")
async def get_property(
    property_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    property_service = PropertyService(db)
    property_obj = property_service.get_property_by_id(property_id)
    if not property_obj:
        raise HTTPException(status_code=404, detail="Property not found")
    return property_obj

@router.post("/")
async def create_property(
    property_data: PropertyCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    property_service = PropertyService(db)
    property_dict = property_data.dict()
    property_dict['agent_id'] = current_user.id
    property_obj = property_service.create_property(property_dict)
    return property_obj

@router.put("/{property_id}")
async def update_property(
    property_id: int,
    property_data: PropertyUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    property_service = PropertyService(db)
    property_dict = {k: v for k, v in property_data.dict().items() if v is not None}
    property_obj = property_service.update_property(property_id, property_dict)
    if not property_obj:
        raise HTTPException(status_code=404, detail="Property not found")
    return property_obj

@router.delete("/{property_id}")
async def delete_property(
    property_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    property_service = PropertyService(db)
    success = property_service.delete_property(property_id)
    if not success:
        raise HTTPException(status_code=404, detail="Property not found")
    return {"message": "Property deleted successfully"}

@router.get("/search/{search_term}")
async def search_properties(
    search_term: str,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    property_service = PropertyService(db)
    properties = property_service.search_properties(search_term)
    return {"properties": properties or []}

@router.get("/{property_id}/analytics")
async def get_property_analytics(
    property_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    property_service = PropertyService(db)
    analytics = property_service.get_property_analytics(property_id)
    if not analytics:
        raise HTTPException(status_code=404, detail="Property not found")
    return analytics

# Image Management Endpoints
@router.get("/{property_id}/images")
async def get_property_images(
    property_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    property_service = PropertyService(db)
    images = property_service.get_property_images(property_id)
    return {"images": images}

@router.post("/{property_id}/images")
async def add_property_image(
    property_id: int,
    image_data: PropertyImageCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    property_service = PropertyService(db)
    image = property_service.add_property_image(property_id, image_data.dict())
    return image

@router.delete("/images/{image_id}")
async def delete_property_image(
    image_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    property_service = PropertyService(db)
    success = property_service.delete_property_image(image_id)
    if not success:
        raise HTTPException(status_code=404, detail="Image not found")
    return {"message": "Image deleted successfully"}

# Document Management Endpoints
@router.get("/{property_id}/documents")
async def get_property_documents(
    property_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    property_service = PropertyService(db)
    documents = property_service.get_property_documents(property_id)
    return {"documents": documents}

@router.post("/{property_id}/documents")
async def add_property_document(
    property_id: int,
    document_data: PropertyDocumentCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    property_service = PropertyService(db)
    document = property_service.add_property_document(property_id, document_data.dict())
    return document

@router.delete("/documents/{document_id}")
async def delete_property_document(
    document_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    property_service = PropertyService(db)
    success = property_service.delete_property_document(document_id)
    if not success:
        raise HTTPException(status_code=404, detail="Document not found")
    return {"message": "Document deleted successfully"}

# Showing Management Endpoints
@router.get("/{property_id}/showings")
async def get_property_showings(
    property_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    property_service = PropertyService(db)
    showings = property_service.get_property_showings(property_id)
    return {"showings": showings}

@router.post("/{property_id}/showings")
async def schedule_property_showing(
    property_id: int,
    showing_data: PropertyShowingCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    property_service = PropertyService(db)
    showing = property_service.schedule_showing(property_id, showing_data.dict())
    return showing

@router.put("/showings/{showing_id}")
async def update_property_showing(
    showing_id: int,
    showing_data: PropertyShowingUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    property_service = PropertyService(db)
    showing = property_service.update_showing(showing_id, showing_data.dict(exclude_unset=True))
    if not showing:
        raise HTTPException(status_code=404, detail="Showing not found")
    return showing

# Bulk Import Endpoint
@router.post("/bulk-import")
async def bulk_import_properties(
    file: UploadFile = File(...),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    property_service = PropertyService(db)
    result = await property_service.bulk_import_properties(file, current_user.id)
    return result

# Property Comparison Endpoint
@router.get("/compare")
async def compare_properties(
    property_ids: str = Query(..., description="Comma-separated property IDs"),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    property_service = PropertyService(db)
    ids = [int(id.strip()) for id in property_ids.split(",")]
    comparison = property_service.compare_properties(ids)
    return comparison

# Map Data Endpoint
@router.get("/map")
async def get_properties_map_data(
    bounds: Optional[str] = Query(None, description="Map bounds: lat1,lng1,lat2,lng2"),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    property_service = PropertyService(db)
    map_data = property_service.get_properties_map_data(bounds)
    return {"properties": map_data}

# Tools Endpoints
@router.post("/{property_id}/cma")
async def generate_cma(
    property_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    property_service = PropertyService(db)
    property_obj = property_service.get_property_by_id(property_id)
    if not property_obj:
        raise HTTPException(status_code=404, detail="Property not found")
    
    cma_data = property_service.generate_cma_report(property_id)
    return cma_data

@router.post("/{property_id}/virtual-tour")
async def create_virtual_tour(
    property_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    property_service = PropertyService(db)
    property_obj = property_service.get_property_by_id(property_id)
    if not property_obj:
        raise HTTPException(status_code=404, detail="Property not found")
    
    tour_data = property_service.create_virtual_tour(property_id)
    return tour_data

@router.post("/{property_id}/market-report")
async def generate_market_report(
    property_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    property_service = PropertyService(db)
    property_obj = property_service.get_property_by_id(property_id)
    if not property_obj:
        raise HTTPException(status_code=404, detail="Property not found")
    
    report_data = property_service.generate_market_report(property_id)
    return report_data

@router.post("/{property_id}/enhance-photos")
async def enhance_photos(
    property_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    property_service = PropertyService(db)
    property_obj = property_service.get_property_by_id(property_id)
    if not property_obj:
        raise HTTPException(status_code=404, detail="Property not found")
    
    enhanced_photos = property_service.enhance_property_photos(property_id)
    return enhanced_photos

@router.post("/{property_id}/optimize-listing")
async def optimize_listing(
    property_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    property_service = PropertyService(db)
    property_obj = property_service.get_property_by_id(property_id)
    if not property_obj:
        raise HTTPException(status_code=404, detail="Property not found")
    
    optimized_listing = property_service.optimize_listing_description(property_id)
    return optimized_listing

@router.post("/{property_id}/social-media-kit")
async def create_social_media_kit(
    property_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    property_service = PropertyService(db)
    property_obj = property_service.get_property_by_id(property_id)
    if not property_obj:
        raise HTTPException(status_code=404, detail="Property not found")
    
    social_kit = property_service.create_social_media_kit(property_id)
    return social_kit

@router.post("/{property_id}/upload-images")
async def upload_property_images(
    property_id: int,
    images: List[UploadFile] = File(...),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    property_service = PropertyService(db)
    property_obj = property_service.get_property_by_id(property_id)
    if not property_obj:
        raise HTTPException(status_code=404, detail="Property not found")
    
    uploaded_images = await property_service.upload_property_images(property_id, images)
    return {"uploaded_count": len(uploaded_images), "images": uploaded_images}

File: estateman-frontend/src/services/properties.ts
import { api } from './api'

export interface Property {
  id: number
  title: string
  description?: string
  property_type: string
  status: string
  price: number
  bedrooms?: number
  bathrooms?: number
  square_feet?: number
  lot_size?: number
  year_built?: number
  address: string
  city: string
  state: string
  zip_code: string
  latitude?: number
  longitude?: number
  agent_id: number
  created_at: string
  updated_at?: string
  images?: PropertyImage[]
  documents?: PropertyDocument[]
  agent?: any
}

export interface PropertyImage {
  id: number
  property_id: number
  image_url: string
  caption?: string
  is_primary: boolean
  order_index: number
}

export interface PropertyDocument {
  id: number
  property_id: number
  document_url: string
  document_name: string
  document_type?: string
  file_size?: number
}

export interface PropertyFilters {
  status?: string
  property_type?: string
  min_price?: number
  max_price?: number
  city?: string
  bedrooms?: number
  bathrooms?: number
}

export interface PropertyCreate {
  address: string
  city: string
  state: string
  zip_code: string
  price: number
  bedrooms?: number
  bathrooms?: number
  square_feet?: number
  property_type: string
  status: string
  description?: string
  agent_id: number
  lot_size?: number
  year_built?: number
  latitude?: number
  longitude?: number
}

export const propertiesService = {
  async getProperties(skip = 0, limit = 100, filters?: PropertyFilters): Promise<Property[]> {
    const params = new URLSearchParams({
      skip: skip.toString(),
      limit: limit.toString(),
      ...Object.fromEntries(
        Object.entries(filters || {}).filter(([_, v]) => v !== undefined).map(([k, v]) => [k, v.toString()])
      )
    })
    
    const response = await api.get<{ properties: Property[] }>(`/properties?${params}`)
    return response.data.properties
  },

  async getProperty(id: number): Promise<Property> {
    const response = await api.get<Property>(`/properties/${id}`)
    return response.data
  },

  async createProperty(propertyData: PropertyCreate): Promise<Property> {
    const response = await api.post<Property>('/properties', propertyData)
    return response.data
  },

  async updateProperty(id: number, propertyData: Partial<PropertyCreate>): Promise<Property> {
    const response = await api.put<Property>(`/properties/${id}`, propertyData)
    return response.data
  },

  async deleteProperty(id: number): Promise<void> {
    await api.delete(`/properties/${id}`)
  },

  async searchProperties(searchTerm: string): Promise<Property[]> {
    const response = await api.get<{ properties: Property[] }>(`/properties/search/${encodeURIComponent(searchTerm)}`)
    return response.data.properties
  },

  async getPropertyAnalytics(id: number): Promise<any> {
    const response = await api.get(`/properties/${id}/analytics`)
    return response.data
  },

  // Image Management
  async getPropertyImages(id: number): Promise<PropertyImage[]> {
    const response = await api.get<{ images: PropertyImage[] }>(`/properties/${id}/images`)
    return response.data.images
  },

  async addPropertyImage(id: number, imageData: { image_url: string; caption?: string; is_primary?: boolean; order_index?: number }): Promise<PropertyImage> {
    const response = await api.post<PropertyImage>(`/properties/${id}/images`, imageData)
    return response.data
  },

  async deletePropertyImage(imageId: number): Promise<void> {
    await api.delete(`/properties/images/${imageId}`)
  },

  // Document Management
  async getPropertyDocuments(id: number): Promise<PropertyDocument[]> {
    const response = await api.get<{ documents: PropertyDocument[] }>(`/properties/${id}/documents`)
    return response.data.documents
  },

  async addPropertyDocument(id: number, documentData: { document_url: string; document_name: string; document_type?: string; file_size?: number }): Promise<PropertyDocument> {
    const response = await api.post<PropertyDocument>(`/properties/${id}/documents`, documentData)
    return response.data
  },

  async deletePropertyDocument(documentId: number): Promise<void> {
    await api.delete(`/properties/documents/${documentId}`)
  },

  // Showing Management
  async getPropertyShowings(id: number): Promise<any[]> {
    const response = await api.get<{ showings: any[] }>(`/properties/${id}/showings`)
    return response.data.showings
  },

  async schedulePropertyShowing(id: number, showingData: { client_name: string; client_email?: string; client_phone?: string; showing_date: string; duration_minutes?: number; notes?: string }): Promise<any> {
    const response = await api.post(`/properties/${id}/showings`, showingData)
    return response.data
  },

  async updatePropertyShowing(showingId: number, showingData: { status?: string; feedback?: string; notes?: string }): Promise<any> {
    const response = await api.put(`/properties/showings/${showingId}`, showingData)
    return response.data
  },

  // Bulk Import
  async bulkImportProperties(file: File): Promise<{ created_count: number; failed_count: number; errors: string[] }> {
    const formData = new FormData()
    formData.append('file', file)
    const response = await api.post('/properties/bulk-import', formData, {
      headers: { 'Content-Type': 'multipart/form-data' }
    })
    return response.data
  },

  // Property Comparison
  async compareProperties(propertyIds: number[]): Promise<{ properties: any[] }> {
    const response = await api.get<{ properties: any[] }>(`/properties/compare?property_ids=${propertyIds.join(',')}`)
    return response.data
  },

  // Map Data
  async getPropertiesMapData(bounds?: string): Promise<Property[]> {
    const params = bounds ? `?bounds=${bounds}` : ''
    const response = await api.get<{ properties: Property[] }>(`/properties/map${params}`)
    return response.data.properties
  },

  // Tools
  async generateCMA(propertyId: number): Promise<any> {
    const response = await api.post(`/properties/${propertyId}/cma`)
    return response.data
  },

  async createVirtualTour(propertyId: number): Promise<any> {
    const response = await api.post(`/properties/${propertyId}/virtual-tour`)
    return response.data
  },

  async generateMarketReport(propertyId: number): Promise<any> {
    const response = await api.post(`/properties/${propertyId}/market-report`)
    return response.data
  },

  async enhancePhotos(propertyId: number): Promise<any> {
    const response = await api.post(`/properties/${propertyId}/enhance-photos`)
    return response.data
  },

  async optimizeListing(propertyId: number): Promise<any> {
    const response = await api.post(`/properties/${propertyId}/optimize-listing`)
    return response.data
  },

  async createSocialMediaKit(propertyId: number): Promise<any> {
    const response = await api.post(`/properties/${propertyId}/social-media-kit`)
    return response.data
  },

  // Image Upload
  async uploadPropertyImages(propertyId: number, images: File[]): Promise<any> {
    const formData = new FormData()
    images.forEach((image, index) => {
      formData.append('images', image)
      formData.append(`is_primary_${index}`, index === 0 ? 'true' : 'false')
    })
    
    const response = await api.post(`/properties/${propertyId}/upload-images`, formData, {
      headers: { 'Content-Type': 'multipart/form-data' }
    })
    return response.data
  }
}